In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("customer-churn-training.csv")

In [10]:
df.head()

,customer_id,tenure_months,support_tickets,monthly_spend_inr,last_login_days,plan_type,churned
0,C001,2,4,499,21,Basic,1
1,C002,18,1,1299,2,Pro,0
2,C003,6,3,799,14,Basic,1
3,C004,30,0,1499,1,Pro,0
4,C005,11,2,999,5,Standard,0


In [4]:
df.shape

(12, 7)

In [11]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 12 entries, 0 to 11
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   customer_id        12 non-null     str  
 1   tenure_months      12 non-null     int64
 2   support_tickets    12 non-null     int64
 3   monthly_spend_inr  12 non-null     int64
 4   last_login_days    12 non-null     int64
 5   plan_type          12 non-null     str  
 6   churned            12 non-null     int64
dtypes: int64(5), str(2)
memory usage: 804.0 bytes


In [ ]:
df.isnull().sum()

In [ ]:
df.duplicated().sum()

In [12]:
df["churned"].value_counts()

churned
0    7
1    5
Name: count, dtype: int64

In [13]:
df["churned"].value_counts(normalize=True) * 100

churned
0    58.333333
1    41.666667
Name: proportion, dtype: float64

In [14]:
majority_class = df["churned"].mode()[0]

print(majority_class)

0


In [17]:
from sklearn.metrics import accuracy_score

print("Scikit-learn is working!")

Scikit-learn is working!


In [18]:
baseline_predictions = [0] * len(df)

baseline_accuracy = accuracy_score(
    df["churned"],
    baseline_predictions
)

print("Baseline Accuracy:", baseline_accuracy)

Baseline Accuracy: 0.5833333333333334


# ML Problem Framing

## Problem
This project aims to predict whether a customer will churn.

## Target Variable
The target variable is `churned`.

- `0` = customer did not churn
- `1` = customer churned

## ML Task
This is a binary classification problem because the target has two possible outcomes: 0 and 1.

## Intended Use
The model may help identify customers who may be at higher risk of churn and help prioritize retention analysis or outreach.

## Prohibited Use
The model should not be used as the sole basis for automatically cancelling services, denying benefits, or making other high-impact decisions about customers.

## Input Features
The planned features are:

- `tenure_months`
- `support_tickets`
- `monthly_spend_inr`
- `last_login_days`
- `plan_type`

The `customer_id` column will not be used as a predictive feature because it is an identifier.

# Responsible Data Card

## 1. Dataset Purpose

This dataset is used to study customer churn and to build a model that predicts whether a customer may churn.

The dataset may support customer retention analysis and prioritization of customers for possible outreach.

The dataset must not be used as the sole basis for high-impact decisions about customers.

## 2. Provenance and Permission

The dataset source and original data collection process are not provided with the dataset.

The identity of the original data creator or owner is not documented.

Consent information and licensing restrictions are not provided.

Therefore, the intended use and permission status should be verified before using this dataset for real-world or production purposes.

## 3. Population and Representation

The dataset contains 12 customer records.

The dataset does not provide demographic or other sensitive-group information, so representation across demographic groups cannot be assessed.

The sampling method is not documented, so it is unknown whether the dataset represents the broader customer population.

The target variable contains:
- 7 customers who did not churn (0)
- 5 customers who churned (1)

Because the dataset is very small, the observed class distribution may not represent the distribution in a larger customer population.

## 4. Features and Target

### Features

- `tenure_months`: Number of months the customer has been with the service.
- `support_tickets`: Number of support tickets associated with the customer.
- `monthly_spend_inr`: Customer's monthly spending in INR.
- `last_login_days`: Number of days since the customer's last login.
- `plan_type`: Customer's subscription plan.

### Target

- `churned`: The outcome we want to predict.
  - `0` = did not churn
  - `1` = churned

### Identifier

- `customer_id`: Customer identifier. It will not be used as a predictive feature.

### Leakage and Sensitive Proxies

No direct target leakage has been identified from the available column descriptions.

However, the dataset is very small, so potential relationships between features and the target should be reviewed carefully.

No sensitive demographic attributes are provided in the dataset. Therefore, sensitive proxies cannot be fully assessed from the available information.

## 5. Quality Checks

### Missing Values
All 7 columns contain 12 non-null values. No missing values were found.

### Duplicates
No duplicate rows were found in the dataset.

### Class Balance
The target variable contains:
- 7 records with `churned = 0`
- 5 records with `churned = 1`

This corresponds to approximately:
- 58.33% non-churned
- 41.67% churned

### Data Types
The dataset contains numerical and categorical features.

### Outliers
No formal outlier analysis has been performed yet.

### Train/Test Separation
A train/test split has not yet been performed. This should be completed before evaluating a trained model.

## 6. Risks and Safeguards

### Bias Risk
The dataset contains only 12 customer records, so the data may not represent the wider customer population.

**Safeguard:** Treat model results as exploratory and evaluate performance on additional representative data before real-world use.

### False Positive Risk
A customer may be predicted as likely to churn even though they would not churn.

**Safeguard:** Predictions should be treated as risk signals rather than certain outcomes.

### False Negative Risk
A customer who is likely to churn may be incorrectly predicted as not likely to churn.

**Safeguard:** Review false negatives during error analysis and consider appropriate evaluation metrics.

### Privacy Risk
The dataset contains customer identifiers.

**Safeguard:** Do not use `customer_id` as a predictive feature and avoid exposing unnecessary customer information.

### Misuse Risk
Predictions could be used to make high-impact decisions about customers.

**Safeguard:** The model should support human review and should not be the sole basis for high-impact customer decisions.

## 7. Intended Evaluation

### Baseline
A majority-class baseline will be used for comparison.

The majority class is `0`, so the baseline predicts `0` for every customer.

### Performance Metrics
The following metrics should be considered:
- Accuracy
- Precision
- Recall
- F1-score

### Error Analysis
False positives and false negatives should be examined to understand where the model makes mistakes.

### Fairness
Fairness evaluation cannot currently be performed because the dataset does not contain demographic or other sensitive-group information.

### Calibration
If the model produces probabilities, calibration should be evaluated before using those probabilities for decision-making.

### Dataset Limitation
Because the dataset contains only 12 records, performance results should be considered educational and exploratory rather than evidence of production-level model performance.

# Risk Register

| Risk | Description | Impact | Mitigation |
|---|---|---|---|
| Small dataset | The dataset contains only 12 customer records. | Model results may be unstable and may not generalize to a larger population. | Treat results as exploratory and validate on a larger dataset before real-world use. |
| Class imbalance | There are 7 non-churned and 5 churned customers. | The model may perform differently across the two classes. | Report class-specific metrics such as precision, recall, and F1-score. |
| False positives | A customer may be predicted as likely to churn when they would not churn. | Unnecessary retention outreach may occur. | Use predictions as risk signals rather than certain outcomes. |
| False negatives | A customer who may churn could be predicted as not likely to churn. | A potentially at-risk customer may be missed. | Review false negatives during error analysis. |
| Customer privacy | The dataset contains a customer identifier. | Customer information could be unnecessarily exposed. | Do not use `customer_id` as a model feature and avoid exposing unnecessary customer information. |
| Data provenance | The source, collection process, consent, and licensing information are not provided. | The dataset may not be appropriate for real-world use without verification. | Verify data ownership, permission, consent, and licensing before production use. |
| Misuse of predictions | Predictions could be treated as certain decisions about customers. | Customers could be unfairly affected. | Keep human review and do not use the model as the sole basis for high-impact decisions. |

In [19]:
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

In [20]:
print("Scikit-learn is working!")

Scikit-learn is working!


In [21]:
X = df[
    [
        "tenure_months",
        "support_tickets",
        "monthly_spend_inr",
        "last_login_days",
        "plan_type"
    ]
]

y = df["churned"]

print("Features shape:", X.shape)
print("Target shape:", y.shape)

Features shape: (12, 5)
Target shape: (12,)


In [22]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.25,
    random_state=42,
    stratify=y
)

print("Training rows:", len(X_train))
print("Testing rows:", len(X_test))

Training rows: 9
Testing rows: 3


In [23]:
preprocessor = ColumnTransformer(
    transformers=[
        (
            "categorical",
            OneHotEncoder(handle_unknown="ignore"),
            ["plan_type"]
        )
    ],
    remainder="passthrough"
)

model = Pipeline(
    steps=[
        ("preprocessor", preprocessor),
        ("classifier", LogisticRegression(max_iter=1000))
    ]
)

In [24]:
model.fit(X_train, y_train)

print("Model training completed!")

Model training completed!


In [25]:
y_pred = model.predict(X_test)

print("Actual values:", y_test.tolist())
print("Predicted values:", y_pred.tolist())

Actual values: [1, 0, 0]
Predicted values: [1, 1, 0]


In [26]:
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, zero_division=0)
recall = recall_score(y_test, y_pred, zero_division=0)
f1 = f1_score(y_test, y_pred, zero_division=0)

print("Model Accuracy:", accuracy)
print("Model Precision:", precision)
print("Model Recall:", recall)
print("Model F1-score:", f1)

Model Accuracy: 0.6666666666666666
Model Precision: 0.5
Model Recall: 1.0
Model F1-score: 0.6666666666666666


In [27]:
baseline_accuracy = accuracy_score(y_test, [0] * len(y_test))

print("Baseline Accuracy:", baseline_accuracy)
print("Model Accuracy:", accuracy)

if accuracy > baseline_accuracy:
    print("The model performed better than the baseline.")
elif accuracy == baseline_accuracy:
    print("The model performed the same as the baseline.")
else:
    print("The model performed worse than the baseline.")

Baseline Accuracy: 0.6666666666666666
Model Accuracy: 0.6666666666666666
The model performed the same as the baseline.


# Model Evaluation and Limitations

## Baseline Comparison

The Logistic Regression model was compared with a majority-class baseline.

The baseline predicts the majority class for every test example, while the Logistic Regression model uses the available customer features.

## Evaluation Metrics

The model was evaluated using:

- Accuracy
- Precision
- Recall
- F1-score

## Limitations

This dataset contains only 12 customer records. The test set contains only 3 records, so the evaluation metrics can change substantially from one data split to another.

Therefore, the results should be treated as educational and exploratory rather than evidence that the model is ready for production use.

A larger and more representative dataset would be needed for reliable evaluation.

## Responsible Use

The model should be treated as a decision-support tool rather than a source of certain predictions. Predictions should not be used as the sole basis for high-impact decisions about customers.